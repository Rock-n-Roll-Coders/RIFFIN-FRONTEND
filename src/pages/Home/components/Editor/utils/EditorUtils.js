import { STEP_COUNT, MIN_BLOCK_COLS, MAX_BLOCK_COLS } from '../EditorConfig'

const calcNewMaxLength = (numberOfCols, numberOfStrings) => (numberOfCols * numberOfStrings) + (numberOfStrings * 2) - 1
const calculateNewFirstCol = (numberOfCols, index) => (numberOfCols + 1) * (index)
const calculateNewLastCol = (numberOfCols, stringNum, index) => (numberOfCols * stringNum) + index

/* 
  Needs an actions object with the following properties:
  area
  characterToAdd
  action
  cols
*/
export const updateBlockValue = (action) => {
  if((action.action === "increase" && action.cols === MAX_BLOCK_COLS) ||
    (action.action === "decrease" && action.cols === MIN_BLOCK_COLS)) {
    return action.area
  }

  const guitarStrings = action.area.split('\n')
  let newValueString = ""
  for(let i = 0; i < guitarStrings.length; i++) {
    const array = [...guitarStrings[i]]
    for(let j = 0; j < STEP_COUNT; j++) {
      if(action.action === "increase") {
        array.push(action.characterToAdd)
      } else {
        array.pop()
      }
    }
    if(i !== guitarStrings.length - 1) {
      array.push("\n")
    }
    newValueString += array.join("")
  }
  return newValueString
}

/* 
  action {
    cols,
    maxLength,
    action,
    stringCount
  }
*/
export const updateTextAreaAttributes = (action) => {
  if((action.action === "increase" && action.cols === MAX_BLOCK_COLS) ||
    (action.action === "decrease" && action.cols === MIN_BLOCK_COLS)) {
    return {
      cols: action.cols,
      maxLength: action.maxLength
    }
  }
  if(action.action === "increase") {
    return {
      cols: action.cols + STEP_COUNT,
      maxLength: calcNewMaxLength(action.cols + STEP_COUNT, action.stringCount)
    }
  } else {   
    return {
      cols: action.cols - STEP_COUNT,
      maxLength: calcNewMaxLength(action.cols - STEP_COUNT, action.stringCount)
    }
  }
}


export const getMapOfLastColumnIndexes = (action) => {
  const newLastCols = {}
  for(let i = 0; i < action.stringCount; i++) {
    const stringNum = i + 1
    const newVal = calculateNewLastCol(action.cols, stringNum, i)
    newLastCols[newVal] = true
  }
  return newLastCols
}

export const getMapOfFirstColumnIndexes = (action) => {
  const newFirstCols = {}
  for(let i = 0; i < action.stringCount; i++) {
    const newVal = calculateNewFirstCol(action.cols, i)
    newFirstCols[newVal] = true
  }
  return newFirstCols
}