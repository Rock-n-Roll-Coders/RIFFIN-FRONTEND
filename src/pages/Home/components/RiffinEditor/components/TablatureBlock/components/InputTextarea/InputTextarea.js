// Components / hooks
import { useRef, useEffect, useState, useContext } from "react";
// Utils / services
import * as utils from "../../../../Utilities";

// MUI
import { useTheme } from "@mui/material/styles";
import { RiffinEditorDispatch } from "pages/Home/components/RiffinEditor/RiffinEditor";
import { DUPLICATION_COLUMN_GAP } from "pages/Home/components/RiffinEditor/EditorConfig";

const InputTextarea = (props) => {
  const [textAreaProperties] = useState({
    cols: props.block.cols,
    // TODO Refactor to numberOfStrings (stringCount is more tangled than youd think >.>)
    stringCount: props.block.numberOfStrings
  })
  const [mapOfLastColumnIndexes, setMapOfLastColumnIndexes] = useState(utils.getMapOfLastColumnIndexes(textAreaProperties))
  const [mapOfFirstColumnIndexes, setMapOfFirstColumnIndexes] = useState(utils.getMapOfFirstColumnIndexes(textAreaProperties))

  const dispatch = useContext(RiffinEditorDispatch);
  const ref = useRef();
  const theme = useTheme();
  
  const inputsStyle = {
    background: "transparent",
    margin: 0,
    fontFamily: "Fira Code",
    position: "relative",
    resize: "none",
    zIndex: 2,
    outline: "none",
    border: "none",
    color: theme.palette.primary.tabInput,
    fontSize: "1.2rem",
    padding: 0
  };

  // ------------- 🏗 MISC HELPERS --------------------------------
  
  /**
   * Checks if the cursor selection is on the last column.
   * @param {Number} selectionStart 
   * @param {Object} mapOfLastColumnIndexes 
   * @returns true if the cursor is on the last column, otherwise false
   */
  const selectionIsOnLastColumn = (selectionStart, mapOfLastColumnIndexes) => (selectionStart in mapOfLastColumnIndexes)

  /**
   * Checks if the cursor selection is on the First column.
   * @param {Number} selectionStart 
   * @param {Object} mapOfFirstColumnIndexes 
   * @returns true if the cursor is on the first column, otherwise false
   */
  const selectionIsOnFirstColumn = (selectionStart, mapOfFirstColumnIndexes) => (selectionStart in mapOfFirstColumnIndexes)

  /**
   * Checks if the cursor selection is on the second to last column. Second to last column is rarely needed and not stored in state.
   * @param {Number} selectionStart 
   * @param {Object} mapOfLastColumnIndexes 
   * @returns true if the cursor is on the second to last column, otherwise false
   */
  const selectionIsOnSecondToLastColumn = (selectionStart, mapOfLastColumnIndexes) => {
    const mapOfSecondToLastColumnIndexes = Object.keys(mapOfLastColumnIndexes).map((val) => val - 1)
    return mapOfSecondToLastColumnIndexes.includes(selectionStart)
  }

  // ------------- 🛠 HELPERS FOR DUPLICATING / DELETING ENTIRE COLUMNS --------------------------------

  /**
   * Gets the positions of the input textarea in a column directly above the curent selection.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @returns an array of positions
   */
  const getPositionsAboveSelection = (selectionStart, cols) => {
    let arrayOfPositions = []
    const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
    for(let position = selectionStart - indexGapBetweenStrings; position >= 0; position -= indexGapBetweenStrings) {
      arrayOfPositions.push(position - 1)
    }
    return arrayOfPositions
  }

  /**
   * Gets the positions of the input textarea in a column directly below the curent selection.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @returns an array of positions
   */
  const getPositionsBelowSelection = (selectionStart, cols, editableLength) => {
    const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
    let arrayOfPositions = []
    for(let position = selectionStart + indexGapBetweenStrings; position <= editableLength; position += indexGapBetweenStrings) {
      arrayOfPositions.push(position - 1)
    }
    return arrayOfPositions
  }
  
  /**
   * Gets the column of positions directly behind the cursor in the input textarea.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @param {Number} maxLength 
   * @param {Number} numberOfStrings 
   * @returns 
   */
  const getColumnOfPositionsBeforeSelection = (selectionStart, cols, maxLength, numberOfStrings) => {
    let numberOfEditablePositionsOnBlock = maxLength - numberOfStrings
    let positionsAboveSelection = getPositionsAboveSelection(selectionStart - 1, cols)
    let positionsBelowSelection = getPositionsBelowSelection(selectionStart - 1, cols, numberOfEditablePositionsOnBlock)
    let allPositionsToDuplicate = [selectionStart - 2, ...positionsAboveSelection, ...positionsBelowSelection]
    return allPositionsToDuplicate
  }

  /**
   * Generates an array of objects represening the inputs to be duplicated and their target positions.
   * @param {Number} selectionStart 
   * @returns an array of objects
   */
  const generateDuplicationValues = (selectionStart) => {
    let positionsToDuplicate = getColumnOfPositionsBeforeSelection(selectionStart, textAreaProperties.cols, props.block.maxLength, textAreaProperties.stringCount)
    let inputsAsArray = [...props.block.inputs]
    const duplicationValues = [];
    positionsToDuplicate.forEach((position) => {
      const characterToDuplicate = inputsAsArray[position];
      duplicationValues.push({
        inputValue: characterToDuplicate,
        dashValue: (characterToDuplicate === " ") ? "-" : " ",
        targetPosition: position + DUPLICATION_COLUMN_GAP
      })
    })
    return duplicationValues;
  }
  // ------------- 🖐 HANDLERS --------------------------------

  /**
   * Sends a dispatch to update the selected block. The new selected block will be whichever block is handling the click.
   * @param {Object} event 
   */
  const handleClick = (event) => {
    const action = {
      type: "updateSelection",
      blockIndex: props.index,
      blockRef: ref,
      selectionStart: event.target.selectionStart
    };
    dispatch(action);
  };

  /**
   * Sends a dispatch to update the selection when an arrow key is pressed. This allows the user to move around the textarea with arrow keys.
   * @param {Object} event 
   */
  const handleKeyUp = (event) => {
    if(utils.isMovementKey(event.key)) {
      const action = {
        type: "updateCursorPosition",
        selectionStart: event.target.selectionStart
      };
      dispatch(action);
    }
  };

  /**
   * Sends a dispatch to insert a character into the textarea. This allows the user to type in the textarea on valid spaces. The only column that is not valid is the last column. This is because there are hidden new-line characters (\n) that keep the structure of the tab intact, making the textarea appear as if it were guitar strings.
   * @param {Number} selectionStart 
   * @param {String} key 
   */
  const handleAddCharacter = (selectionStart, key) => {
    let action = {};
    if(selectionIsOnLastColumn(selectionStart - 1, mapOfLastColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
     action ={ 
        type: "addCharacter",
        character: key,
        selectionStart: selectionStart - 1
      };
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to remove a character into the textarea. This allows the user to delete on valid spaces. The only column that is not valid is the first column. This prevents the user form deleting a new-line characters (\n) on the previous space, which keeps the structure of the tab intact.
   * @param {*} selectionStart 
   */
  const handleDeleteCharacter = (selectionStart) => {
    let action = {};
    if(selectionIsOnFirstColumn(selectionStart + 1, mapOfFirstColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart + 1
      };
    } else {
      action = {
        type: "deleteCharacter",
        selectionStart: selectionStart
      };
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to duplicate an entire column in the textarea. The values of the duplicated column will be pasted two spaces over, leaving one blank space between the original and duplicated columns.
   * @param {*} selectionStart 
   */
  const handleDuplicateColumn = (selectionStart) => {
    let action = {};
    if(selectionIsOnLastColumn(selectionStart - 1, mapOfLastColumnIndexes)
      || selectionIsOnFirstColumn(selectionStart - 1, mapOfFirstColumnIndexes)
      || selectionIsOnSecondToLastColumn(selectionStart - 1, mapOfLastColumnIndexes)
    ) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
      action = {
        type: "duplicateColumn",
        duplicationValues: generateDuplicationValues(selectionStart),
        selectionStart: selectionStart
      }
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to delete an entire column in the textarea.
   * @param {Number} selectionStart 
   */
  const handleDeleteColumn = (selectionStart) => {
    let action = {};
    if(selectionIsOnFirstColumn(selectionStart - 1, mapOfFirstColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
      const positionsToDelete = getColumnOfPositionsBeforeSelection(selectionStart, textAreaProperties.cols, props.block.maxLength, textAreaProperties.stringCount)
      action = {
        type: 'deleteColumn',
        positionsToDelete,
        selectionStart: selectionStart - 1
      }
    }
    dispatch(action);
  }

  /**
   * Checks the pressed key and routes to the appropriate dispatch handler.
   * @param {Object} event 
   */
  const handleChange = (event) => {
    event.preventDefault();
    const key = event.nativeEvent.data || "Backspace";
    const dispatchType = utils.getDispatchTypeOfPressedKey(key);
    if(dispatchType === undefined) {
      let action = {
        type: 'updateCursorPosition',
        selectionStart: event.target.selectionStart - 1
      };
      dispatch(action);
    } else if(dispatchType === "addCharacter") {
      handleAddCharacter(event.target.selectionStart, key)
    } else if(dispatchType === "deleteCharacter") {
      handleDeleteCharacter(event.target.selectionStart)
    } else if(dispatchType === "duplicateColumn") {
      handleDuplicateColumn(event.target.selectionStart)
    } else if(dispatchType === "deleteColumn") {
      handleDeleteColumn(event.target.selectionStart)
    }
  };

  // ------------- 🦾 EFFECTS --------------------------------
  
  /**
   * Updates the bounry maps when the size of the block changes.
   */
  useEffect(() => {
    setMapOfLastColumnIndexes(utils.getMapOfLastColumnIndexes(textAreaProperties))
    setMapOfFirstColumnIndexes(utils.getMapOfFirstColumnIndexes(textAreaProperties))
  }, [textAreaProperties, props.block.cols]);

  return (
    <textarea
      style={inputsStyle}
      value={props.block.inputs}
      onChange={handleChange}
      onKeyUp={handleKeyUp}
      onPaste={(event) => event.preventDefault()}
      onClick={handleClick}
      cols={textAreaProperties.cols}
      rows={textAreaProperties.stringCount}
      maxLength={props.block.maxLength}
      ref={ref}
    />
  );
}
 
export default InputTextarea;