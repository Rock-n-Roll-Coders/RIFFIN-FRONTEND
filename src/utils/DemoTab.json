{
    "_id": "demo",
    "owner": "",
    "numberOfStrings": 6,
    "name": "Riffin Blues (demo tab)",
    "blocks": [
        {
            "label": "Riffin Editor core features:\n- Change the size of the staff using the Staff Menu ===>\n\n- Take advantage of the Duplicate Chord feature by pressing \"]\". This will duplicate an entire chord starting at the position directly behind the cursor.\n\n- Rearrange staffs by dragging the handle (=) on the left edge.\n\n- Above each staff is an area for notes!\n\n- Catalog your tablature with Tags. Tags can be added to the bar at the top of the screen.\n\nSAMPLE TAB: \nTuning: EADGBE\n\n\n\nGuitar 1:\nRepeat twice for 8 bars - add some bluesy flare to that strummin!",
            "inputs": "|                                      |\n|                                      |\n|                                      |\n|                                      |\n| 2 2 4 4  2 2 4 4  2 2 4 4  2 2 4 4   |\n| 0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0   |",
            "dashes": "|--------------------------------------|\n|--------------------------------------|\n|--------------------------------------|\n|--------------------------------------|\n|- - - - -- - - - -- - - - -- - - - ---|\n|- - - - -- - - - -- - - - -- - - - ---|",
            "cols": 40,
            "maxLength": 251,
            "_id": "637bdb2aa557ebd717828c1e"
        },
        {
            "label": "Guitar 1:\nNow throw in 4 bars of that sweet open A power chord.\nGotta throw in some bluesy rhythm!",
            "inputs": "|                                      |\n|                                      |\n|                                      |\n| 2 2 4 4  2 2 4 4  2 2 4 4 5 5 4 4    |\n| 0 0 0 0  0 0 0 0  0 0 0 0 0 0 0 0    |\n|                                      |",
            "dashes": "|--------------------------------------|\n|--------------------------------------|\n|--------------------------------------|\n|- - - - -- - - - -- - - - - - - - ----|\n|- - - - -- - - - -- - - - - - - - ----|\n|--------------------------------------|",
            "cols": 40,
            "maxLength": 251,
            "_id": "637be4bd6065d9183d028a91"
        },
        {
            "label": "Guitar 1:\nBack to the start again... ",
            "inputs": "|                                      |\n|                                      |\n|                                      |\n|                                      |\n| 2 2 4 4  2 2 4 4  2 2 4 4  2 2 4 4   |\n| 0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0   |",
            "dashes": "|--------------------------------------|\n|--------------------------------------|\n|--------------------------------------|\n|--------------------------------------|\n|- - - - -- - - - -- - - - -- - - - ---|\n|- - - - -- - - - -- - - - -- - - - ---|",
            "cols": 40,
            "maxLength": 251,
            "_id": "637bdb2aa557ebd717828c1e"
        },
        {
            "label": "Guitar 1:\nWrap up with a bluesy walk-back!",
            "inputs": "|                                                          1~~  |\n|                                                          0~~  |\n|                                                          2~~  |\n|  4 4 4 4 4 4 2~~ 2 2 2 2 2                               1~~  |\n|  2 2 2 2 2 2 0~~ 0 0 0 0 0 2 2 5 5 5 4 4 4 3 3 3 2 0h1h2 2~~  |\n|                            0 0 0 0 0 0 0 0 0 0 0 0            |",
            "dashes": " ----------------------------------------------------------   --|\n|----------------------------------------------------------   --|\n|----------------------------------------------------------   --|\n|-- - - - - - -   - - - - - -------------------------------   --|\n|-- - - - - - -   - - - - - - - - - - - - - - - - - -     -   --|\n|---------------------------- - - - - - - - - - - - ------------|",
            "cols": 65,
            "maxLength": 401,
            "_id": "637be4bd6065d9183d028a93"
        },
        {
            "label": "Guitar 2:\nCome in swinging with some blues licks!",
            "inputs": "|                                                     |\n|   14 / 17 17 17 17 17 17 17 15 17 15                |\n|                                      14/ 16         |\n|                                             14  14~~|\n|                                                     |\n|                                                     |",
            "dashes": "|-----------------------------------------------------|\n|---  - -  -  -  -  -  -  -  -  -  -  ----------------|\n|--------------------------------------   -  ---------|\n|---------------------------------------------  --    |\n|-----------------------------------------------------|\n|-----------------------------------------------------|",
            "cols": 55,
            "maxLength": 341,
            "_id": "637be4bd6065d9183d028a94"
        }
    ],
    "tags": [
        "Blues",
        "ðŸŽ¸",
        "Need to practice"
    ]
}