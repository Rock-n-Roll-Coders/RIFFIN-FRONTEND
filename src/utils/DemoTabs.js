export const demoTabs = [
    {
        "owner": "",
        "numberOfStrings": 6,
        "name": "12 bar blues",
        "blocks": [
            {
                "label": "TUNING: EADGBE\n\nBluesy intro walk",
                "inputs": "|                          1 1 0  |\n|                          0 0 0  |\n|                          2 2    |\n|                          1 1    |\n| 7 7 5 5 4 4 3 3 2 0 1 2  2 2    |\n|                                 |",
                "dashes": "|-------------------------- - - --|\n|-------------------------- - - --|\n|-------------------------- - ----|\n|-------------------------- - ----|\n|- - - - - - - - - - - - -- - ----|\n|---------------------------------|",
                "cols": 35,
                "maxLength": 221,
                "_id": "6382930bff50a5d590725d89",
                "createdAt": "2022-11-26T22:28:27.447Z",
                "updatedAt": "2022-11-26T22:28:27.447Z"
            },
            {
                "label": " ",
                "inputs": "|                                                                              |\n|                                                                              |\n|                                                                              |\n|                     2 2 4 4  2 2 4 4                                         |\n| 2 2 4 4   2 2 4 4   0 0 0 0  0 0 0 0   2 2 4 4  2 2 4 4  2 2 4 4  2 2 4 4    |\n| 0 0 0 0   0 0 0 0                      0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0    |",
                "dashes": "|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|\n|--------------------- - - - -- - - - -----------------------------------------|\n|- - - - --- - - - --- - - - -- - - - --- - - - -- - - - -- - - - -- - - - ----|\n|- - - - --- - - - ---------------------- - - - -- - - - -- - - - -- - - - ----|",
                "cols": 80,
                "maxLength": 491,
                "_id": "6382930bff50a5d590725d8a",
                "createdAt": "2022-11-26T22:28:27.448Z",
                "updatedAt": "2022-11-26T22:28:27.448Z"
            },
            {
                "label": "  ",
                "inputs": "|                                                                              |\n|                                                                              |\n|                                                                              |\n| 2 2 4 4  2 2 4 4  2 2 4 4  2 2 4 4                                           |\n| 0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0  2 2 4 4  2 2 4 4  2 2 4 4   2 2 4 4      |\n|                                     0 0 0 0  0 0 0 0  0 0 0 0   0 0 0 0      |",
                "dashes": "|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|\n|- - - - -- - - - -- - - - -- - - - -------------------------------------------|\n|- - - - -- - - - -- - - - -- - - - -- - - - -- - - - -- - - - --- - - - ------|\n|------------------------------------- - - - -- - - - -- - - - --- - - - ------|",
                "cols": 80,
                "maxLength": 491,
                "_id": "6382930bff50a5d590725d8b",
                "createdAt": "2022-11-26T22:28:27.448Z",
                "updatedAt": "2022-11-26T22:28:27.448Z"
            },
            {
                "label": " ",
                "inputs": "|                                        |                                     |\n|                                        |   then repeat from intro walk       |\n|                                        |                                     |\n| 4 4 6 6  4 4 6 6  2 2 4 4  2 2 4 2     |                                     |\n| 2 2 2 2  2 2 2 2  0 0 0 0  0 0 0 0 2   |                                     |\n|                                    0 0 |                                     |",
                "dashes": "|---------------------------------------- -------------------------------------|\n|---------------------------------------- ---    -      -    -     -    -------|\n|---------------------------------------- -------------------------------------|\n|- - - - -- - - - -- - - - -- - - - ----- -------------------------------------|\n|- - - - -- - - - -- - - - -- - - - - --- -------------------------------------|\n|------------------------------------ - - -------------------------------------|",
                "cols": 80,
                "maxLength": 491,
                "_id": "6382930bff50a5d590725d8c",
                "createdAt": "2022-11-26T22:28:27.448Z",
                "updatedAt": "2022-11-26T22:28:27.448Z"
            }
        ],
        "tags": [
            "blues",
            "practice"
        ],
        "_id": "demo_blues_riff",
        "createdAt": "2022-11-26T22:28:27.448Z",
        "updatedAt": "2022-11-26T22:28:27.448Z",
        "__v": 0
    },
    {
      "_id": "riffin_sketchpad_tutorial",
      "owner": "c52209cb-6e0c-49f9-a4b0-cb5dbd25ae58",
      "numberOfStrings": 6,
      "name": "Riffin sketchpad tutorial",
      "blocks": [
          {
              "label": "Click the 'EDIT' button to open this tab in the sketchpad. The sketchpad is only available on large displays, if you're viewing this on mobile you will not see an 'EDIT' button.\n===\n",
              "inputs": "|    0            0          |\n|    0          0   0        |\n|    1        1       1      |\n|    2      2           2    |\n|    2    2               2  |\n|    0  0                   0|",
              "dashes": "|---- ------------ ----------|\n|---- ---------- --- --------|\n|---- -------- ------- ------|\n|---- ------ ----------- ----|\n|---- ---- --------------- --|\n|---- -- ------------------- |",
              "cols": 30,
              "maxLength": 191,
              "_id": "637c3c40c682ca4f60eb9341"
          },
          {
              "tempKey": "Wed Nov 23 2022 20:10:46 GMT-0500 (Eastern Standard Time)0.6697429678666145",
              "inputs": "|                                                |\n|                                                |\n|                                                |\n|                                                |\n|                                                |\n|                                                |",
              "dashes": "|------------------------------------------------|\n|------------------------------------------------|\n|------------------------------------------------|\n|------------------------------------------------|\n|------------------------------------------------|\n|------------------------------------------------|",
              "cols": 50,
              "maxLength": 311,
              "label": "Below is a staff for tablature. Give yourself some space to type - use the \"Staff Menu\" on the right to increase the length, then click anywhere on the staff to start writing! \n\nThe arrow keys are helpful for navigating."
          },
          {
              "label": "Above each staff is an area for notes. Take as many lines as you need.",
              "inputs": "|                                                15/17 17^ (bend all the way!) |\n|                                                                              |\n|   You can write notes on the staff too!                                      |\n|                                                                              |\n|                                                                              |\n|                                                                              |",
              "dashes": "|------------------------------------------------     -   -     -   -   -     -|\n|------------------------------------------------------------------------------|\n|---   -   -     -     -  -   -     -    --------------------------------------|\n|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|\n|------------------------------------------------------------------------------|",
              "cols": 80,
              "maxLength": 491,
              "_id": "637c54d29e1c0a4310a8f3bc"
          },
          {
              "label": "Duplicate a chord behind the cursor by pressing \"]\". Try duplicating the D chord below!\nCheck out the 'Special inputs' panel for more cool tricks.\n",
              "inputs": "| 0| <--chord behind cursor:  0 0           D chord ->  2                      |\n| 0                           0 0                       4                      |\n| 1                           1 1                       2                      |\n| 2                           2 2                       0                      |\n| 2                           2 2                       x                      |\n| 0                           0 0                       x                      |",
              "dashes": "|-  -        -      -       -- - ----------- -     -  -- ----------------------|\n|- --------------------------- - ----------------------- ----------------------|\n|- --------------------------- - ----------------------- ----------------------|\n|- --------------------------- - ----------------------- ----------------------|\n|- --------------------------- - ----------------------- ----------------------|\n|- --------------------------- - ----------------------- ----------------------|",
              "cols": 80,
              "maxLength": 491,
              "_id": "637c54d29e1c0a4310a8f3bd"
          },
          {
              "label": "Use the 'add tag' input at the top of the screen to help organize your tablature.",
              "inputs": "|                                                |\n|                                                |\n<-- use the drag handle (=) to rearrange staffs! |\n|                                                |\n|                                                |\n|                                                |",
              "dashes": "|------------------------------------------------|\n|------------------------------------------------|\n   -   -   -    -      -   -  -         -       -|\n ------------------------------------------------|\n|------------------------------------------------|\n|------------------------------------------------|",
              "cols": 50,
              "maxLength": 311,
              "_id": "637c58127707340387e4d856"
          }
      ],
      "tags": [
          "Practice"
      ]
  }
]