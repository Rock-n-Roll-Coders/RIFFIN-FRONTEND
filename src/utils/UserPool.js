import { CognitoUserPool } from "amazon-cognito-identity-js";
// UserPoolId comes from the general tab in Cognito
// ClientID comes from the App Client Settings in Cognito
const poolData = {
  UserPoolId: process.env.REACT_APP_COGNITO_USER_POOL_ID,
  ClientId: process.env.REACT_APP_COGNITO_CLIENT_ID,
};

export default new CognitoUserPool(poolData);
