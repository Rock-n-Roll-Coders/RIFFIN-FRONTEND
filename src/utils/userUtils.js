const getIdTokenFromUser = (user) => {
  let dataKey = user.userDataKey;
  dataKey = dataKey.replace(".userData", ".idToken");
  const idToken = user.storage[dataKey];
  return idToken;
};

export { getIdTokenFromUser };
