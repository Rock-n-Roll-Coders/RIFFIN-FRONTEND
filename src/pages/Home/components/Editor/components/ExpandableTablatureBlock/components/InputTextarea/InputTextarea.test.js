import { getNewGuitarBlock } from "../../../../utils/EditorUtils";
import InputTextarea from "./InputTextarea";
import renderer from 'react-test-renderer';
// Note that style props may be undefined undefined because I'm not wrapping this in the theme context provider.

test('renders correctly using a new tablatureBlock', ()=> {
  const block = getNewGuitarBlock();
  const handleBlockChange = jest.fn()
  const handleKeyUpInBlock = jest.fn()
  const handleClickedBlock = jest.fn()
  const index = 0
  const tree = renderer
    .create(<InputTextarea 
      block={block}
      handleBlockChange={handleBlockChange}
      handleKeyUpInBlock={handleKeyUpInBlock}
      handleClickedBlock={handleClickedBlock}
      index={index}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});