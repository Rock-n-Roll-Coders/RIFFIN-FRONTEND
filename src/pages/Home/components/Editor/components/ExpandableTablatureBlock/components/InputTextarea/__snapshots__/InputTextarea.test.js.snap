// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly using a new tablatureBlock 1`] = `
<textarea
  cols={40}
  maxLength={251}
  onChange={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": "none",
      "color": undefined,
      "fontFamily": "Fira Code",
      "fontSize": "1.2rem",
      "margin": 0,
      "outline": "none",
      "padding": 0,
      "position": "relative",
      "resize": "none",
      "zIndex": 2,
    }
  }
  value="|                                      |
|                                      |
|                                      |
|                                      |
|                                      |
|                                      |"
/>
`;
